Disable fee estimation in blocksonly mode (by removing the fee estimates global)

darosior opened the pull request:
If the `blocksonly` mode is turned on after running with transaction
relay enabled for a while, the fee estimation will serve outdated data
to both the internal wallet and to external applications that might be
feerate-sensitive and make use of `estimatesmartfee` (for example a
Lightning Network node).

This has already caused issues (for example https://github.com/bitcoin/bitcoin/issues/16840 (C-lightning), or https://github.com/lightningnetwork/lnd/issues/2562 (LND)) and it seems prudent to fail rather than to give inaccurate values.

This fixes #16840, and closes #16890 which tried to fix the symptoms (RPC) but not the cause as mentioned by sdaftuar :
> If this is a substantial problem, then I would think we should take action to protect our own wallet users as well (rather than hide the results of what our fee estimation would do!).

promag commented on src/init.cpp:
```diff
@@ -1750,8 +1750,10 @@ bool AppInitMain(NodeContext& node)
+    // Allowed to fail as this file IS missing on first startup. Also, don't initialize
+    // fee estimation with old data if we don't relay transactions, as they would never be
+    // updated.
+    if (!est_filein.IsNull() && g_relay_txes)
```
Why read the file then? You could early check like:
```cpp
// Don't initialize fee estimation if we don't relay transactions.
if (!g_relay_txes) {
    fs::path est_path = GetDataDir() / FEE_ESTIMATES_FILENAME;
    CAutoFile est_filein(fsbridge::fopen(est_path, "rb"), SER_DISK, CLIENT_VERSION);
    if (!est_filein.IsNull()) ::feeEstimator.Read(est_filein);
    fFeeEstimatesInitialized = true;
}
```
This also avoids writing to estimates file in `Shutdown`.

darosior replied:
I think
```cpp
fFeeEstimatesInitialized = true;
```
should be set anyway ?

What do you think about this conceptually ?

promag replied:
I think `fFeeEstimatesInitialized` exists to handle early shutdowns. But if you don't read estimates then writing will clear it right?

darosior replied:
Hmm right

darosior commented:
@promag : addressed your feedback. Would be interested about what do you think about disabling the fee estimation in `blocksonly` conceptually though.

luke-jr commented:
Concept ACK / Sounds reasonable.

brakmic commented:
Concept ACK.

laanwj commented:
Concept ACK, but I don't know enough about the underlying mechanism to be sure if this is a reasonable way to accomplish it.

promag commented:
> Would be interested about what do you think about disabling the fee estimation in `blocksonly` conceptually though.

Same opinion as above.

MarcoFalke commented:
Concept ACK, while the fee estimation should work for some limited time after the mempool has been turned off (see https://github.com/bitcoin/bitcoin/issues/16840#issuecomment-529904220), I can see the point to fail early on startup instead of later at runtime.

MarcoFalke commented:
Concept ACK 33ca35902435a162c526bf6c8bd82145585dfbd9

darosior commented:
@sdaftuar (ping because you reviewed the previous PR) : what do you think of this approach compared to the previous one ?

@morcos (ping because you implemented the fee estimation) : what do you think of it implementation-wise ?

darosior committed:
Remove the remaining fee estimation globals

This moves the CBlockPolicyEstimator to the NodeContext, which get rids
of two globals and allows us to conditionally create the
CBlockPolicyEstimator (and to remove a circular dep).

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior commented:
This was intended to be a "quick fix", but the approach seems to be too much of a workaround.

I've reworked this to opt for a more involved approach based on #19556. This now:
- Removes the fee estimator related globals.
- Conditionally creates the fee estimator: if we don't relay transactions we don't even bother to allocate the `CBlockPolicyEstimator`.

This is rebased on top of #19556 (which i itself rebased on top of #19604).

darosior committed:
init: don't create a CBlockPolicyEstimator if we don't relay transactions

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior commented:
Rebased on  #19556 now that the remove-mempool-global series is almost over.

jonatack commented:
Concept ACK, will review after 19556 goes in.

jnewbery commented:
Concept ACK

jnewbery commented:
Intuitively, I don't think the mempool should 'own' the fee estimator, but rather I see the fee estimator as a client of the mempool. I'd suggest constructing the fee estimator in `AppInitMain()` and owning the memory in `NodeContext`, like the other components.

darosior commented:
> Intuitively, I don't think the mempool should 'own' the fee estimator, but rather I see the fee estimator as a client of the mempool. I'd suggest constructing the fee estimator in AppInitMain() and owning the memory in NodeContext, like the other components.

It seems more intuitive to me for the mempool to own the fee estimator as the latter can not exist without the former, and it would otherwise be a potentially-uninitialized globally-accessible "second class" object ? I think of this as "the mempool *may* provide fee estimation".

jnewbery commented:
> it would otherwise be a potentially-uninitialized globally-accessible "second class" object

That depends on how you write the code. `NodeContext` holds (mostly) unique ptrs which may or may not be initialized. It's up to the components to make sure that they're using a component that is enabled (see, for example, the usage of banman in net_processing).

Fee estimator is currently a separate component from the mempool. Let's not add unnecessary coupling between the components by moving it inside the mempool. Decoupled components with well defined interfaces make it easier to reason about and test the code.

darosior commented:
Refactored to keep the feeEstimator as a separate component as per @jnewbery request (spoiler: it's cleaner now).

jnewbery commented on src/txmempool.h:
```diff
@@ -577,6 +578,9 @@ class CTxMemPool
 
     uint64_t CalculateDescendantMaximum(txiter entry) const EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    void initFeeEstimator(CAutoFile& est_file);
```
This is unused.

darosior replied:
Argh, leftover from the refactor..

jnewbery commented on src/txmempool.h:
```diff
@@ -20,6 +20,7 @@
 #include <optional.h>
 #include <policy/feerate.h>
 #include <primitives/transaction.h>
+#include <streams.h>
```
Why is this added?

darosior replied:
Same, was for CAutoFile

jnewbery commented on src/init.cpp:
```diff
@@ -235,18 +234,18 @@ void Shutdown(NodeContext& node)
     }
 
-    if (fFeeEstimatesInitialized)
+    if (node.feeEstimator)
     {
```
Join lines to fix up with current code style:

`if (node.feeEstimator) {`

(see https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c)

jnewbery commented on src/init.cpp:
```diff
@@ -235,18 +234,18 @@ void Shutdown(NodeContext& node)
         CAutoFile est_fileout(fsbridge::fopen(est_path, "wb"), SER_DISK, CLIENT_VERSION);
         if (!est_fileout.IsNull())
-            ::feeEstimator.Write(est_fileout);
+            node.feeEstimator->Write(est_fileout);
```
Again, fix up braces when touching adjacent lines.

jnewbery commented on src/init.cpp:
```diff
@@ -1356,15 +1355,27 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     // is not yet setup and may end up being set up twice if we
     // need to reindex later.
 
+    // see Step 2: parameter interactions for more information about these
```
I think only `g_relay_txes` needs to be moved up here.

darosior replied:
Yeah but it made sense to keep them all grouped imho

jnewbery commented on src/init.cpp:
```diff
@@ -1768,9 +1774,8 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
-    if (!est_filein.IsNull())
-        ::feeEstimator.Read(est_filein);
-    fFeeEstimatesInitialized = true;
+    if (node.feeEstimator && !est_filein.IsNull())
```
use braces for if block.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -327,11 +327,13 @@ class ChainImpl : public Chain
     CFeeRate estimateSmartFee(int num_blocks, bool conservative, FeeCalculation* calc) override
     {
-        return ::feeEstimator.estimateSmartFee(num_blocks, calc, conservative);
+        if (!m_node.feeEstimator) return CFeeRate(0);
```
I think you can just use default initialization of `CFeeRate`:

```suggestion
        if (!m_node.feeEstimator) return {};
```

jnewbery commented on src/node/context.h:
```diff
@@ -36,6 +37,7 @@ class WalletClient;
 struct NodeContext {
     std::unique_ptr<CConnman> connman;
     std::unique_ptr<CTxMemPool> mempool;
+    std::shared_ptr<CBlockPolicyEstimator> feeEstimator;
```
Why does this need to be shared? `NodeContext` should have unique ownership of the feeEstimator object.

darosior replied:
Right, fixed

jnewbery commented on src/rpc/mining.cpp:
```diff
@@ -1064,7 +1064,13 @@ static RPCHelpMan estimatesmartfee()
     RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
-    unsigned int max_target = ::feeEstimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+
+    NodeContext& node = EnsureNodeContext(request.context);
+    if (!node.feeEstimator) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Fee estimation disabled");
+    }
```
How about extracting this to an `EnsureFeeEstimator()` function? (see `EnsureMemPool` and `EnsureChainman`)

darosior replied:
I thought `EnsureXX()` were used for components that should not be absent, while the fee estimator may be ?

jnewbery replied:
I think it's fine to add `EnsureXX()` for components that may not be enabled. Does that sound right to you @MarcoFalke ?

MarcoFalke replied:
It doesn't make sense to call `estimatesmartfee` without the fee estimator. So yes, `EnsureFeeEstimator` is the way to go right now. (In the future the fee estimator can register the RPC itself, so when it is missing, the RPC won't even be there)

darosior replied:
> In the future the fee estimator can register the RPC itself, so when it is missing, the RPC won't even be there

Hmm that would be particularly nice, as for example applications up the stack relying on fee estimation would be able to just error at startup rather than "check and see if it errors". Might propose a follow-up implementing this.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -331,7 +331,7 @@ void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee,
 }
 
-CTxMemPool::CTxMemPool(CBlockPolicyEstimator* estimator)
+CTxMemPool::CTxMemPool(std::shared_ptr<CBlockPolicyEstimator> estimator)
```
You don't need to make this change. The fee estimator's memory/lifetime is owned by `NodeContext`. This can continue being a raw pointer to that object (in the same way that `PeerManager` holds pointers and references to other components).

Not making this change should reduce the diff of this PR considerably.

jnewbery commented on src/init.cpp:
```diff
@@ -1356,15 +1355,27 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    // Don't initialize fee estimation with old data if we don't relay transactions,
+    // as they would never get updated.
+    assert(!node.feeEstimator);
+    if (g_relay_txes) {
```
Consider joining into one line for simple one-statement if blocks:

```suggestion
    if (g_relay_txes) node.feeEstimator = std::make_shared<CBlockPolicyEstimator>();
```

jnewbery commented on src/init.cpp:
```diff
@@ -1356,15 +1355,27 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    // as they would never get updated.
+    assert(!node.feeEstimator);
+    if (g_relay_txes) {
+        node.feeEstimator = std::make_shared<CBlockPolicyEstimator>();
```
You're constructing the fee estimator here. You should destruct it on shutdown. At line 304 (after the `node.mempool` unique ptr is reset), you should reset this ptr too.

darosior replied:
Dropped this somehow.. Fixing now.

jnewbery commented on test/functional/feature_fee_estimation.py:
```diff
@@ -263,6 +264,13 @@ def run_test(self):
+        self.log.info("Testing that fee estimation is disabled in blocksonly.")
+        self.stop_nodes()
+        self.start_nodes(extra_args=[["-blocksonly"]] * len(self.nodes))
+        for node in self.nodes:
```
You only need to test this on one node.

darosior replied:
Thanks, amended to only start and test on one node.

jnewbery commented on src/init.cpp:
```diff
@@ -1768,10 +1774,9 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     // Allowed to fail as this file IS missing on first startup.
-    if (!est_filein.IsNull())
-        ::feeEstimator.Read(est_filein);
-    fFeeEstimatesInitialized = true;
```
I think this flag might protect us from overwriting the estimates.dat file in case we abort start-up for some reason. Rather than removing it entirely, I think it should be moved to CBlockPolicyEstimator, and then checked on destruction.

If you really wanted to encaptulate this better, you could move this file read logic to the feeEstimator constructor, and the flush logic to a public flush method in CBlockPolicyEstimator, and hide all the construction/destruction logic from init.cpp.

darosior replied:
Good catch. Went for the encapsulation way, will push a commit.

jnewbery replied:
Moving this code to the destructor may be a behaviour change, since there will have been changes to the mempool in the intervening period. I suggest adding a public flush method and calling it here to avoid the behaviour change.

darosior replied:
Yeah, i initially swapped the two calls to `reset()`, but as you prefer.

MarcoFalke commented on src/node/context.h:
```diff
@@ -36,6 +37,7 @@ class WalletClient;
 struct NodeContext {
     std::unique_ptr<CConnman> connman;
     std::unique_ptr<CTxMemPool> mempool;
+    std::unique_ptr<CBlockPolicyEstimator> feeEstimator;
```
```suggestion
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
```
for new code

MarcoFalke commented on src/rpc/mining.cpp:
```diff
@@ -1064,7 +1064,10 @@ static RPCHelpMan estimatesmartfee()
     RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
-    unsigned int max_target = ::feeEstimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+
+    CBlockPolicyEstimator& feeEstimator = EnsureFeeEstimator(request.context);
```
```suggestion
    CBlockPolicyEstimator& fee_estimator = EnsureFeeEstimator(request.context);
```

snake case for new code ;)

darosior replied:
Thanks! Amended

maflcko reviewed (commented):
Concept ACK

darosior commented:
Amended to use snake case, and added a commit to encapsulate the management of the fee estimation file in `CBlockPolicyEstimator`.

jnewbery commented on src/policy/fees.cpp:
```diff
@@ -886,8 +902,14 @@ bool CBlockPolicyEstimator::Write(CAutoFile& fileout) const
 
-bool CBlockPolicyEstimator::Read(CAutoFile& filein)
-{
+bool CBlockPolicyEstimator::ReadEstFile() {
```
You don't need a separate function for this since it's only called in one place. You can just insert this code into the constructor.

jnewbery commented on src/init.cpp:
```diff
@@ -1356,15 +1344,27 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    // Don't initialize fee estimation with old data if we don't relay transactions,
+    // as they would never get updated.
+    if (g_relay_txes) {
+        fs::path est_path = GetDataDir() / FEE_ESTIMATES_FILENAME;
```
Can you move this line inside the constructor and avoid changing the signature of the constructor?

darosior replied:
Thought it would be better to keep the filename constants in init ?..

jnewbery replied:
No need. If it's only used in policy/fees.cpp you can move the constant there.

darosior replied:
Ok, done.

jnewbery commented:
This is very confusing. In your latest branch you're now calling `FlushUnconfirmed()` twice, which is definitely a behaviour change. I don't think you need to change the dtor at all, or the interface for the ctor.

See https://github.com/jnewbery/bitcoin/tree/pr18766.1. The only thing that has to change is a new `flush()` function is added and some logic is moved inside the `CBlockPolicyEstimator` class.

darosior commented:
> This is very confusing. In your latest branch you're now calling FlushUnconfirmed() twice, which is definitely a behaviour change. I don't think you need to change the dtor at all, or the interface for the ctor.

I pushed in a hurry yesterday, and noticed this afterwards. Will fix this and the rest as soon as possible, but want to take some time for that to avoid wasting your time (and potentially others' too) again by pushing half-baked patches. Coming back to it soon (:tm:)

darosior committed:
feestimator: encapsulate estimation file logic

This moves the fee_estimates file management to the CBlockPolicyEstimator
Flush() method.

Co-authored-by: John Newbery <john@johnnewbery.com>
Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior commented:
Ok, it should be in a reviewable state now. This:
- Moves the fee estimator from a global to the node context.
- Encapsulates the `fee_estimates.dat` logic in CBlockPolicyEstimator.
- Avoid creating the fee estimator at all if we don't relay transactions.

jnewbery commented on src/policy/fees.cpp:
```diff
@@ -504,10 +507,27 @@ CBlockPolicyEstimator::CBlockPolicyEstimator()
+    if (est_filepath.empty()) return;
+
+    CAutoFile est_file(fsbridge::fopen(est_filepath, "wb"), SER_DISK, CLIENT_VERSION);
+    if (est_file.IsNull() || !Write(est_file)) {
```
Why are you moving this to the destructor? Are we sure that nothing happens to the fee estimator's state between  https://github.com/bitcoin/bitcoin/pull/18766/files#diff-c865a8939105e6350a50af02766291b7R236 and https://github.com/bitcoin/bitcoin/pull/18766/files#diff-c865a8939105e6350a50af02766291b7R294?

darosior replied:
Right. What do you think of instead exposing a `RecordEstimation()` method to the `CBlockPolicyEstimator` ?

jnewbery replied:
Was there a reason you didn't like the flush() method in https://github.com/jnewbery/bitcoin/tree/pr18766.1?

darosior replied:
Ok didn't see this. Had a version with both the FlushUnconfirmed() and RecordEstimation() separated locally. Grouping them and pushing in a minute !

jnewbery commented:
utACK 4105c63a67a40fe109b2f2c020c9fbf25d1fc250

If you retouch this, you can remove the policy/fees.h #include in validation.cpp:

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 86d18a9d7f..7d60dd6e20 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -22,7 +22,6 @@
 #include <logging/timer.h>
 #include <node/ui_interface.h>
 #include <optional.h>
-#include <policy/fees.h>
 #include <policy/policy.h>
 #include <policy/settings.h>
 #include <pow.h>
```

fjahr commented:
Code review ACK 4105c63a67a40fe109b2f2c020c9fbf25d1fc250

If you retouch you could also check the description of 4105c63a67a40fe109b2f2c020c9fbf25d1fc250, which saying logic has move to the destructor while it is actually in the `Flush()` method now.

jnewbery commented:
Oh, one more thing - you could make the git commit logs just a little more verbose. No need to write an essay, but just a short justification for the changes.

darosior commented:
Yeah, did not want to invalidate the ACKs but now it didn't make it in before the feature freeze, i'll address both your comments.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
Le mercredi, octobre 21, 2020 11:16 AM, John Newbery <notifications@github.com> a écrit :

> Oh, one more thing - you could make the git commit logs just a little more verbose. No need to write an essay, but just a short justification for the changes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/18766#issuecomment-713430502), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFLK3F7VQKCDVYV26ZFRYSDSL2RFNANCNFSM4MQ2ROMA).

jnewbery commented on src/init.cpp:
```diff
@@ -1384,13 +1384,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     assert(!node.connman);
     node.connman = MakeUnique<CConnman>(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), args.GetBoolArg("-networkactive", true));
 
-    assert(!node.fee_estimator);
```
I think you should keep this assert.

darosior replied:
Good catch, will re-add it if i have to retouch.

jnewbery replied:
Just re-add it now. I'm the only person who's ACKed the latest head and I'm happy to reACK.

darosior replied:
Ok, done.

jnewbery replied:
Thanks!

darosior commented:
Rebased. Addressed @jnewbery and @fjahr nits.

jnewbery reviewed (commented):
utACK 4105c63a67a40fe109b2f2c020c9fbf25d1fc250

jnewbery commented:
utACK 4aaad74c4c8a24c4b67af4e92c29805b923f2225

fjahr commented:
Code review ACK 4aaad74c4c8a24c4b67af4e92c29805b923f2225

ellemouton commented:
Hi @darosior :) quick question: Will serving outdated fee data not also be a problem if the node has been off for a while and the fee estimation files are just stale in general? i.e would it not perhaps be useful to look at 'last-updated' time on the data and then starting fresh if the data is too old?

darosior commented:
Hi @ellemouton, fortunately we discard these estimations already. If the node has been off for a while, we'll exponentially [decay the estimation as we connect new blocks](https://github.com/bitcoin/bitcoin/blob/42b66a6b814bca130a9ccf0a3f747cf33d628232/src/policy/fees.cpp#L575-L622) :).

ellemouton commented:
Aha! That makes sense. Thanks for the context @darosior 😄

MarcoFalke commented on src/init.cpp:
```diff
@@ -1389,10 +1389,13 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     node.connman = MakeUnique<CConnman>(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), args.GetBoolArg("-networkactive", true));
 
+    assert(!node.fee_estimator);
+    node.fee_estimator = MakeUnique<CBlockPolicyEstimator>();
```
cc21167415aa01db862035cd07b6b9b35fb53f35

would be nice if new code used `std::make_unique` instead of the deprecated wrapper

darosior replied:
Included, had to rebase anyway. Used it for CTxMemPool too as it was free diff-wise.

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -141,7 +142,9 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
     pblocktree.reset(new CBlockTreeDB(1 << 20, true));
 
-    m_node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
+    m_node.fee_estimator = MakeUnique<CBlockPolicyEstimator>();
```
same

MarcoFalke commented on src/init.cpp:
```diff
@@ -236,18 +235,18 @@ void Shutdown(NodeContext& node)
-    if (fFeeEstimatesInitialized)
-    {
-        ::feeEstimator.FlushUnconfirmed();
+    if (node.feeEstimator) {
```
cc21167415aa01db862035cd07b6b9b35fb53f35

```
  CXX      libbitcoin_server_a-init.o
init.cpp:238:14: error: no member named 'feeEstimator' in 'NodeContext'; did you mean 'fee_estimator'?
    if (node.feeEstimator) {
             ^~~~~~~~~~~~
             fee_estimator
./node/context.h:40:44: note: 'fee_estimator' declared here
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
                                           ^
init.cpp:239:14: error: no member named 'feeEstimator' in 'NodeContext'; did you mean 'fee_estimator'?
        node.feeEstimator->FlushUnconfirmed();
             ^~~~~~~~~~~~
             fee_estimator
./node/context.h:40:44: note: 'fee_estimator' declared here
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
                                           ^
init.cpp:243:18: error: no member named 'feeEstimator' in 'NodeContext'; did you mean 'fee_estimator'?
            node.feeEstimator->Write(est_fileout);
                 ^~~~~~~~~~~~
                 fee_estimator
./node/context.h:40:44: note: 'fee_estimator' declared here
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
                                           ^
init.cpp:1801:14: error: no member named 'feeEstimator' in 'NodeContext'; did you mean 'fee_estimator'?
    if (node.feeEstimator && !est_filein.IsNull()) {
             ^~~~~~~~~~~~
             fee_estimator
./node/context.h:40:44: note: 'fee_estimator' declared here
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
                                           ^
init.cpp:1802:14: error: no member named 'feeEstimator' in 'NodeContext'; did you mean 'fee_estimator'?
        node.feeEstimator->Read(est_filein);
             ^~~~~~~~~~~~
             fee_estimator
./node/context.h:40:44: note: 'fee_estimator' declared here
    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
                                           ^
5 errors generated.

darosior replied:
Fixed..

MarcoFalke commented on src/interfaces/node.cpp:
```diff
@@ -207,8 +207,9 @@ class NodeImpl : public Node
+        if (!m_context->fee_estimator) return {};
         FeeCalculation fee_calc;
-        CFeeRate result = ::feeEstimator.estimateSmartFee(num_blocks, &fee_calc, conservative);
+        CFeeRate result = m_context->fee_estimator->estimateSmartFee(num_blocks, &fee_calc, conservative);
```
cc21167415aa01db862035cd07b6b9b35fb53f35

this is unused. Instead of shuffling code around, you may add a commit that removes it.

```diff
diff --git a/src/interfaces/node.cpp b/src/interfaces/node.cpp
index 2c5f8627e6..a9ca4f9cf8 100644
--- a/src/interfaces/node.cpp
+++ b/src/interfaces/node.cpp
@@ -205,15 +205,6 @@ public:
         }
     }
     bool getNetworkActive() override { return m_context->connman && m_context->connman->GetNetworkActive(); }
-    CFeeRate estimateSmartFee(int num_blocks, bool conservative, int* returned_target = nullptr) override
-    {
-        FeeCalculation fee_calc;
-        CFeeRate result = ::feeEstimator.estimateSmartFee(num_blocks, &fee_calc, conservative);
-        if (returned_target) {
-            *returned_target = fee_calc.returnedTarget;
-        }
-        return result;
-    }
     CFeeRate getDustRelayFee() override { return ::dustRelayFee; }
     UniValue executeRpc(const std::string& command, const UniValue& params, const std::string& uri) override
     {
diff --git a/src/interfaces/node.h b/src/interfaces/node.h
index 5079be038e..36f76aeb4f 100644
--- a/src/interfaces/node.h
+++ b/src/interfaces/node.h
@@ -151,9 +151,6 @@ public:
     //! Get network active.
     virtual bool getNetworkActive() = 0;
 
-    //! Estimate smart fee.
-    virtual CFeeRate estimateSmartFee(int num_blocks, bool conservative, int* returned_target = nullptr) = 0;
-
     //! Get dust relay fee.
     virtual CFeeRate getDustRelayFee() = 0;

darosior replied:
Right, i didn't notice.. Cherry-picked your patch and adapted it to the new interface path post-rebase.
```patch
Author: Antoine Poinsot <darosior@protonmail.com>
Date:   Tue Dec 1 11:24:33 2020 +0100

    interface: remove unused estimateSmartFee method from node
    
    Co-Authored-by: MarcoFalke <falke.marco@gmail.com>
    Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

diff --git a/src/interfaces/node.h b/src/interfaces/node.h
index 5079be038..36f76aeb4 100644
--- a/src/interfaces/node.h
+++ b/src/interfaces/node.h
@@ -151,9 +151,6 @@ public:
     //! Get network active.
     virtual bool getNetworkActive() = 0;
 
-    //! Estimate smart fee.
-    virtual CFeeRate estimateSmartFee(int num_blocks, bool conservative, int* returned_target = nullptr) = 0;
-
     //! Get dust relay fee.
     virtual CFeeRate getDustRelayFee() = 0;
 
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index 77a5957a5..3cd9c6fda 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -221,15 +221,6 @@ public:
         }
     }
     bool getNetworkActive() override { return m_context->connman && m_context->connman->GetNetworkActive(); }
-    CFeeRate estimateSmartFee(int num_blocks, bool conservative, int* returned_target = nullptr) override
-    {
-        FeeCalculation fee_calc;
-        CFeeRate result = ::feeEstimator.estimateSmartFee(num_blocks, &fee_calc, conservative);
-        if (returned_target) {
-            *returned_target = fee_calc.returnedTarget;
-        }
-        return result;
-    }
     CFeeRate getDustRelayFee() override { return ::dustRelayFee; }
     UniValue executeRpc(const std::string& command, const UniValue& params, const std::string& uri) override
     {
```

maflcko reviewed (changes requested):
Concept ACK cc21167415aa01db862035cd07b6b9b35fb53f35  🚟

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK cc21167415aa01db862035cd07b6b9b35fb53f35  🚟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUilvAv/d0DZXa+aEjrm/fPzFOmqe/7bGZZTctNR8VQxeQ8DE+Ft7Q7zWo4bvHYl
YeB42SZeTofhU5+ICOGsmxytQ52FA0bPvOJXGH/8zQ6wIvrdTxAhjt5MP3YunOy9
mt8qOOMtLi/RybMV4VgEPDqYZRz27s6LuUxgbgK8YJ5q981pp15LhQ5e0Au4duU0
MQfqwZIqV7L7H5DEcLo/VeN4hVurII5vLg3PK1a56APtW2MKi4lPx5kS17TuAZY5
12kew9JuoaGVrAJKHHzXme0Q+yCwOgptZ1rVEhEYsB11aRG9kkThXvO+oo+oPtc6
SdcGtilVdnfE6x/VNEiD7VTxWgbQ5BUUAPoAoWIrOMDsWB15rcw/dtlAXoHF8JFR
pOJUCK4racs+TPjfwGtAVpjBmqiebzZFNIRPjizMDCKmxEnMCLNNX2l3jR3U3x8M
ObOSwoIIF8W384rupBCjF2+aVnmOsJ2ktx8AFdoNX1NdAs81f0NIfOyoW4hfOl4N
7l5DRdLQ
=6uOX
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `b08551e34b42f9bb8f8e283ece496e30ae7d418c171ecf636076b60c63a83df6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108b08551e34b42f9bb8f8e283ece496e30ae7d418c171ecf636076b60c63a83df6f010f996cab2f6444a6430a36218bce0ef1808fff0103afb12fdd3d11a8f8172617a8985c47608f1045fc5fea9f008417b91381bd6ed530083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01065bba345e05126e8702a579a33efc07808f1045fc5fea9f008791dcc80583fcf870083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d3a11a0529d0da161083da590d33c3e408f1045fc5fea8f008eed629e751d7583b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ede3a820685ed4ff533841572d408bb408f1045fc5feacf008bf2da9fce9b522ef0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

darosior committed:
interface: remove unused estimateSmartFee method from node

Co-Authored-by: MarcoFalke <falke.marco@gmail.com>
Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior commented on src/init.cpp:
```diff
@@ -1795,10 +1798,9 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     // Allowed to fail as this file IS missing on first startup.
-    if (!est_filein.IsNull())
-        ::feeEstimator.Read(est_filein);
-    fFeeEstimatesInitialized = true;
```
Right, removed the "refactor" prefix from the commit.

jnewbery commented:
utACK 01e39989f23b26a23cea26e5ec4d8dfc2010bc97

MarcoFalke commented on src/policy/fees.cpp:
```diff
@@ -856,6 +868,17 @@ CFeeRate CBlockPolicyEstimator::estimateSmartFee(int confTarget, FeeCalculation
+
+    CAutoFile est_file(fsbridge::fopen(est_filepath, "wb"), SER_DISK, CLIENT_VERSION);
+    if (est_file.IsNull() || !Write(est_file)) {
+        LogPrintf("%s: Failed to write fee estimates to %s\n", __func__, est_filepath.string());
```
same

darosior replied:
Removed

maflcko reviewed (approved):
left some style nits. Let me know if you want to fix them or get this merged instead.

review ACK 01e39989f23b26a23cea26e5ec4d8dfc2010bc97 🐪

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 01e39989f23b26a23cea26e5ec4d8dfc2010bc97 🐪
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhJMgwAwnGF/r7ezo/CimEvIuOfeD6LzJ7NDb6SHs5vwyofUN2w8UojfZ7MxuQV
9kGtWH4oHtzNU5732r2HYPfAGWZI6iJe8HAnZY5fjzC0KoN8PLY2DJ97lS9/rWnb
5Dco0RD4r+g0ZVFnN+mfMrQ47nmDHCiULNP0nbav40iXlWS8IQiZiNNJVKInW0q7
w5M3e3I1Np2PQ3lEMqMJ4An2Dt+WAtHLeA9OYB4b0yoXIc4JtYiiQ/mrogEqkAv/
tWal6S7ElJGmuL+/ceXNg6xXqvKpPT6U4yNxcKR9s/ceyhEngaOt7EmDWWRVVXt7
PkV29eLM4TYmPbasK6dIUei2q+xtDWjzKFgAuMe83+kWDyb8+oqNxZi140xHVG2P
AYlbfKD4JnoGB32tEfzgh/Z9dQj4xG+mWhLYL8I7oUDU/SkC3blBmpwPiSHxmvA8
NWabgQshHcDeKWiw3GpZs6BbMjI6xJnq9Dq8kOmYX5Hz6uT3mIJ76KN31WRHFUfC
6kbMbc1m
=5+en
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `532e63b9bee4dd7e9da189eb2890bcc77d9c1f37ba300a2a322c9d027b0c9158  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108532e63b9bee4dd7e9da189eb2890bcc77d9c1f37ba300a2a322c9d027b0c9158f010b4e5666b4004cff7694e64df4e8f6fb808fff0100cef337cac7fa2cc2fb31682e04831ba08f1200da09a97dad003363140273e3215ee24236af4aecc750313294c5bda8926d38d08f1045fc7b6fdf00804b22628bb83492b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010436554d8b6822f0d6ba0c80bfe1d2c3708f1045fc7b6fdf008dd2e4dfcb7ee80df0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

darosior commented on test/functional/feature_fee_estimation.py:
```diff
@@ -262,6 +263,13 @@ def run_test(self):
+        self.log.info("Testing that fee estimation is disabled in blocksonly.")
+        self.stop_nodes()
+        self.nodes[0].start(extra_args=["-blocksonly"])
+        self.nodes[0].wait_for_rpc_connection()
```
Done, thanks

darosior commented on src/policy/fees.cpp:
```diff
@@ -500,6 +503,15 @@ CBlockPolicyEstimator::CBlockPolicyEstimator()
+
+    // If the fee estimation file is present, read recorded estimations
+    fs::path est_filepath = GetDataDir() / FEE_ESTIMATES_FILENAME;
+    if (est_filepath.empty()) return;
```
Well, there is a comment just above.. If the file isn't present (likely first start), don't report a failure to read fee estimates: it's expected. Kind of the echo of the `// Allowed to fail as this file IS missing on first startup.` [comment in the red part of the diff](https://github.com/bitcoin/bitcoin/pull/18766/files/8527183fde72d097700da18600eb7dc37c1ba206..01e39989f23b26a23cea26e5ec4d8dfc2010bc97#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dL1795-L1798).

Thanks, you however ringed a bell in my head: this check should *not* be present in Flush or the file would never be created!

MarcoFalke replied:
what I mean is that a path representation that is (not) empty (https://en.cppreference.com/w/cpp/filesystem/path/empty) doesn't imply anything about the presence of a file.

You can think of fs::path as a "fancy string". empty means the string is empty. non-empty means the string is non-empty, but it doesn't imply the file exists (or even any of the parent folders)

darosior replied:
Thanks.. That was clearly stupid..
Removed as i don't think i can use C++17's `fs::exists()` instead?

darosior commented on src/policy/fees.cpp:
```diff
@@ -500,6 +503,15 @@ CBlockPolicyEstimator::CBlockPolicyEstimator()
+
+    CAutoFile est_file(fsbridge::fopen(est_filepath, "rb"), SER_DISK, CLIENT_VERSION);
+    if (est_file.IsNull() || !Read(est_file)) {
+        LogPrintf("%s: Failed to read fee estimates from %s\n", __func__, est_filepath.string());
```
> is normally hit on first start.

It's not, on first start we would have returned just above.

Removed the `__func__`.

MarcoFalke replied:
> Also the log message could be suffixed with ". Continue anyway.\n" to clarify that this failure does not hinder program execution and is normally hit on first start.

Has this been addressed or responded to?

darosior replied:
No, this was missed. Done now.

darosior commented:
Addressed all @MarcoFalke 's remaining nits.

practicalswift commented:
Concept ACK

MarcoFalke commented:
re-ACK 4e28753f60 👋
only changes:
* Small test fixup (restart_node)
* Small log fixups
* Remove dead code (empty path check)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 4e28753f60 👋
only changes:
* Small test fixup (restart_node)
* Small log fixups
* Remove dead code (empty path check)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiivwwAvP/ngY1HjLgcIUIL6koFzy3aWPRMQhVCyB4gj6n7rHQoETxZZKMrgk4Z
soRCvS8dCX1B9OMgJ7B4iBcyX2ywfw7nAA4PwUSPMJLbNitqzPkpxuG1RNLtLfwt
4loau6a5ZZJlKYhLoR8Q6NEdYwCC/OfpZYRIjlQd2+waGdaki1/G86Q6LKGe5xAZ
TM4VW89RpnHofXm5zxJZq3CiUExZMMSGVMwp+4eGpk0h9/eeYoKr1IGFh7qGzTSe
X5U/t5z6EmygD3gL1QaVWSb73TTeaU2Iux3U7JR9pYJgK4dNjZr3WKTCY7NJORqv
kCcW3ZHL4E3w8Oz9hnU6uSHsj+xOONTRxffHzsP7e1nkYTRSqEjqAohO1u2jL0Ho
2FSyXyDtp2uC0mn64JW0duVZI88FT6fRqyt/HiUHvnFzLSdTg90mDMtWMF/qQcb2
jXpEvjaKGI129CgdlHRxVpUFmAK6bOdHxrx9fgcwD1L6hHo1PQX8N59ulLP2uano
g2unQ65X
=AMhQ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3d9e3dc48bc4d73872a862f7eccfe39d2af942484475d2f8af25b52833aa2031  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083d9e3dc48bc4d73872a862f7eccfe39d2af942484475d2f8af25b52833aa2031f01085bbf7ed73edc39d6b0270c6467f6ae908fff01023189a9538d482ef1509eac479de34e908f1045fcddfd3f0083d753d5612d257010083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b5f0af50342e1c06a6c2cedc7ed14c0408f1045fcddfd3f008df58306c2d6616f50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010c4e18b40b114f76de75adea74e6825cf08f1045fcddfd3f008800afeb339ab296c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d2443f3163d2e6e09e16762a63ffdff808f1045fcddfd3f008b0e533220428eaf80083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

jnewbery commented:
utACK 4e28753f60613ecd35cdef87bef5f99c302c3fbd
